algorithmic-strategies/dynamic-programming.md

```diff
@@ -0,0 +1,302 @@
+| ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
+| Dynamic Programming | An algorithmic strategy of breaking a problem down into subproblems that need to be calculated multiple times, allowing us to improve performance by storing results and reusing them. | "This algorithm calculates the same subproblem over and over, so let's use a dynamic-programming approach to improve the performance!" |
+| Memoization         | An optimization technique used primarily to speed up algorithms by storing the results of subproblems and returning the cached result when the same subproblem occurs again.           | "If we memoize the results of that function call, we can improve the performance if it gets called again with the same arguments."     |
+
+## Example: Fibonacci
+
+The classic Fibonacci sequence is inherently recursive, but also inefficient to solve using the typical recursive manner.
+
+In the Fibonacci sequence `fibonacci(n)`:
+
+`fibonacci(0)` = 0, for \(n = 0\)  
+`fibonacci(1)` = 1, for \(n = 1\)  
+`fibonacci(n)` = `fibonacci(n-1)` + `fibonacci(n-2)`, for \(n > 1\)
+
+We could code the Fibonacci sequence as follows:
+
+```python
+def fibonacci(n):
+    if n == 0 or n == 1:
+        return n
+
+    return fibonacci(n - 1) + fibonacci(n - 2)
+```
+
+However this is wildly inefficient. Note below for the Fibonacci of 5, how `fibonacci(2)` is called 3 times, and `fibonacci(1)` is called 5 times. As \(n\) grows larger, this will occur more and more often. For any \(n > 1\), we end up making \(2^n\) method calls!
+
+![fibonacci(5) calls fibonacci(4)(a) and fibonacci(3)(b). fibonacci(4)(a) calls fibonacci(3)(c) and fibonacci(2)(d). fibonacci(3)(c) calls fibonacci(2)(e) and fibonacci(1)(f). fibonacci(2)(e) calls fibonacci(1)(g) and fibonacci(0)(h). fibonacci(2)(d) calls fibonacci(1)(i) and fibonacci(0)(j). fibonacci(3)(b) calls fibonacci(2)(k) and fibonacci(1)(l). fibonacci(2)(k) calls fibonacci(1)(m) and fibonacci(0)(n).](../assets/algorithmic-strategies_dynamic-programming_fibonacci-inefficient.png)  
+_Fig. Fibonacci of 5. Notice how the same subproblems are repeatedly called!_
+
+Instead of solving the same problems over and over again we can solve these problems by storing them in a _memo_ and using the stored subproblems to make calculating the larger problem more efficient.
+
+Let's take a look at a few variations of `fibonacci` which make use of memoization to improve their performance.
+
+<br />
+
+<details>
+
+<summary>Forward iterative solution. Starts from the base cases of [0, 1] building up to the desired number.</summary>
```

Comment:
By "forward," I meant starting from 0 and calculating up to the final solution, as opposed to the initial recursive solution, which starts at the end, and works back to bottom of the recursion. I didn't really need to add the term to the iterative solution, other than that it parallels the first of the dynamic programming recursive solutions. So really it just turn into a handy label that I could use to refer to these a little later on.

Open to other suggestions. Maybe I could change the order to more like:

```suggestion
<summary>Iterative solution that starts from the base cases of [0, 1] building up to the desired number. Because it starts from the lower Fibonacci numbers and counts up, we can refer to this as a forward iterative solution.</summary>
```
or maybe
```suggestion
<summary>Forward iterative solution. We call it a forward solution since it starts from the base cases of [0, 1] and then builds up to the desired result.</summary>
```

Reply:
But maybe for this one, simplest is best?

```suggestion
<summary>Iterative solution. Starts from the base cases of [0, 1] building up to the desired number.</summary>
```